pkg/container/queue/chunkqueue.go

```diff
@@ -0,0 +1,389 @@
+
+package queue
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/pingcap/log"
+)
+
+const (
+	// the size of each chunk is 1024 bytes (1kB) by default
+	defaultSizePerChunk = 1024
+	// the minimum length of each chunk is 16
+	minimumChunkLen      = 16
+	defaultPitchArrayLen = 16
+)
+
+// ChunkQueue is a generic, efficient, iterable and GC-friendly queue.
+// Attention, it's not thread-safe.
+type ChunkQueue[T any] struct {
+	// [head, tail) is the section of chunks in use
+	head int
+	tail int
+
+	// size is number of elements in queue
+	size int
+
+	// chunks is an array storing ptr
+	chunks []*chunk[T]
+	// chunkLength is the max number of elements stored in every chunk
+	chunkLength  int
+	chunkPool    sync.Pool
+	defaultValue T
+}
+
+func (q *ChunkQueue[T]) firstChunk() *chunk[T] {
+	return q.chunks[q.head]
+}
+
+func (q *ChunkQueue[T]) lastChunk() *chunk[T] {
+	return q.chunks[q.tail-1]
+}
+
+// NewChunkQueue creates a new ChunkQueue
+func NewChunkQueue[T any]() *ChunkQueue[T] {
+	return NewChunkQueueLeastCapacity[T](1)
+}
+
+// NewChunkQueueLeastCapacity creates a ChunkQueue with an argument minCapacity.
+// It requests that the queue capacity be at least minCapacity. And it's similar
+// to the cap argument when making a slice using make([]T, len, cap)
+func NewChunkQueueLeastCapacity[T any](minCapacity int) *ChunkQueue[T] {
+	elementSize := unsafe.Sizeof(*new(T))
+	if elementSize == 0 {
+		log.Error("Cannot create a queue of type")
+		return nil
+	}
+
+	chunkLength := int(defaultSizePerChunk / elementSize)
+	if chunkLength < minimumChunkLen {
+		chunkLength = minimumChunkLen
+	}
+
+	q := &ChunkQueue[T]{
+		head:        0,
+		tail:        0,
+		size:        0,
+		chunkLength: chunkLength,
+	}
+	q.chunkPool = sync.Pool{
+		New: func() any {
+			return newChunk[T](q.chunkLength, q)
+		},
+	}
+
+	q.chunks = make([]*chunk[T], defaultPitchArrayLen, defaultPitchArrayLen)
+	q.extend(minCapacity)
+	return q
+}
+
+// Size returns the number of elements in queue
+func (q *ChunkQueue[T]) Size() int {
+	return q.size
+}
+
+// Len returns the number of elements in queue
+func (q *ChunkQueue[T]) Len() int {
+	return q.size
+}
+
+// Cap returns the capacity of the queue. The queue can hold more elements
+// than that number by automatic expansion
+func (q *ChunkQueue[T]) Cap() int {
+	return q.chunkLength*(q.tail-q.head) - q.chunks[q.head].l
+}
+
+// Empty indicates whether the queue is empty
+func (q *ChunkQueue[T]) Empty() bool {
+	return q.size == 0
+}
+
+// At returns the value of a given index. At() does NOT support modifying the value
+func (q *ChunkQueue[T]) At(idx int) (T, bool) {
+	if idx < 0 || idx >= q.size {
+		return q.defaultValue, false
+	}
+	i := q.chunks[q.head].l + idx
+	return q.chunks[q.head+i/q.chunkLength].data[i%q.chunkLength], true
+}
+
+// Replace assigns a new value to a given index
+func (q *ChunkQueue[T]) Replace(idx int, val T) bool {
+	if idx < 0 || idx >= q.size {
+		return false
+	}
+	i := q.chunks[q.head].l + idx
```

Comment:
Why not use `firstChunk` here? Did you want to explicitly state anything? Maybe we add some comments to explain what happened here?

Reply:
1)Yes, and 2)No. There are some checks in the previous version of `firstChunk()` and I didn't use it here. For now, it makes sense to replace it.